File: callbacks/Demo3.java

interface Callback<T> {
    void onComplete(T success, String failure);
}

class Demo3 {

    static String getString() {
        try {
            Thread.sleep(2000);
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        return "Hello World!";
    }

    static int getInteger() {
        try {
            Thread.sleep(2000);
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        return 123;
    }

    static <T> void doApiCall(int x, Callback<T> listener, Class<T> type) {
        if (x == 0) {
            String response = getString();

            try  {
                listener.onComplete(type.cast(response), null);
            } catch (Exception e) {
                listener.onComplete(null, "Failed to cast response.");
            }
        } else {
            int response = getInteger();

            try {
                listener.onComplete(type.cast(response), null);
            } catch (Exception e) {
                listener.onComplete(null, "Failed to cast response.");
            }
        }
    }

    public static void main(String[] args) {
        doApiCall(0, (success, failure) -> {
		    if (success != null) System.out.println("Success " + success);
		    else System.out.println("Failure " + failure);
		}, String.class);

        doApiCall(0, (success, failure) -> {
		    if (success != null) System.out.println("Success " + success);
		    else System.out.println("Failure " + failure);
		}, Integer.class);

        doApiCall(1, (success, failure) -> {
            if (success != null) System.out.println("Success " + success);
            else System.out.println("Failure " + failure);
        }, Integer.class);

        doApiCall(1, (success, failure) -> {
            if (success != null) System.out.println("Success " + success);
            else System.out.println("Failure " + failure);
        }, String.class);

    }
}

File: callbacks/Demo3.kt
fun getString(): String {
    try {
        Thread.sleep(2000)
    } catch (e: Exception) {
        println(e.message)
    }

    return "Hello World!"
}

fun getInteger(): Int {
    try {
        Thread.sleep(2000)
    } catch (e: Exception) {
        println(e.message)
    }

    return 123
}

inline fun <reified T> doApiCall(
    x: Int,
    onComplete: (success: T?, failure: String?) -> Unit
) {
    if (x == 0) {
        val response = getString()

        try {
            onComplete(response as T, null)
        } catch (e: Exception) {
            onComplete(null, "Failed to cast response.")
        }
    } else {
        val response = getInteger()

        try {
            onComplete(response as T, null)
        } catch (e: Exception) {
            onComplete(null, "Failed to cast response.")
        }
    }
}

fun main() {
    doApiCall<String>(0, { success, failure ->
        success?.let { println(it) }
        failure?.let { println(it) }
    })

    doApiCall<Int>(0, { success, failure ->
        success?.let { println(it) }
        failure?.let { println(it) }
    })


    doApiCall<Int>(1, { success, failure ->
        success?.let { println(it) }
        failure?.let { println(it) }
    })

    doApiCall<String>(1, { success, failure ->
        success?.let { println(it) }
        failure?.let { println(it) }
    })

}
